lib/TPP/Transforms/LoopInsertion.cpp

```diff
@@ -0,0 +1,293 @@
+#include "TPP/Passes.h.inc"
+} // namespace tpp
+} // namespace mlir
+
+using namespace mlir;
+using namespace mlir::scf;
+
+namespace mlir {
+namespace tpp {
+
+static SmallVector<ReassociationIndices>
+getReassociationIndices(ArrayRef<int64_t> origtensorShape,
+                        SmallVector<SmallVector<unsigned>> tileShapes) {
+  SmallVector<ReassociationIndices> indices;
+
+  size_t index = 0;
+  for (size_t i = 0; i < tileShapes.size(); i++) {
+    ReassociationIndices reassociationIndex;
+    for (size_t j = 0; j < tileShapes[i].size(); j++)
+      reassociationIndex.push_back(index++);
+    indices.push_back(reassociationIndex);
+  }
+  for (size_t i = tileShapes.size(); i < origtensorShape.size(); i++) {
+    ReassociationIndices reassociationIndex;
+    reassociationIndex.push_back(index++);
+    indices.push_back(reassociationIndex);
+  }
+
+  return indices;
+}
+
+void insertSubview(ArrayRef<int64_t> tensorShape, Type type, Type resultType,
+                   SmallVector<ReassociationIndices> reassociation,
+                   Value operand, ForallOp op, OpBuilder b,
+                   xsmm::BrgemmOp brgemmOp, int operandNumber) {
+  b.setInsertionPoint(op);
+  auto expandShape = b.create<memref::ExpandShapeOp>(
+      op.getLoc(),
+      MemRefType::get({tensorShape},
+                      dyn_cast<MemRefType>(type).getElementType()),
+      operand, reassociation);
+  expandShape.setStaticOutputShape(tensorShape);
+  b.setInsertionPoint(&op.getBody()->front());
+  SmallVector<OpFoldResult> strides(tensorShape.size(), b.getIndexAttr(1)),
+      sizes, offsets;
+  size_t tileSize =
+      tensorShape.size() - dyn_cast<ShapedType>(resultType).getShape().size();
+  SmallVector<int64_t> tileSizes;
+  for (size_t i = 0; i < tensorShape.size(); i++) {
+    if (i < tileSize) {
+      int opnum = operandNumber;
+      if (opnum == 3) {
+        opnum = 1;
+      }
+      int inductionVarIndex = (opnum - 1) * tileSize + i;
+
+      offsets.push_back(op.getInductionVars()[inductionVarIndex]);
+      sizes.push_back(b.getIndexAttr(1));
+    } else {
+      sizes.push_back(b.getIndexAttr(tensorShape[i]));
+      tileSizes.push_back(tensorShape[i]);
+      offsets.push_back(b.getIndexAttr(0));
+    }
+  }
+
+  auto subviewType =
+      MemRefType::get({tileSizes}, dyn_cast<MemRefType>(type).getElementType());
+  auto [originalStride, originalOffset] =
+      getStridesAndOffset(dyn_cast<MemRefType>(subviewType));
+  subviewType = MemRefType::get(
+      {tileSizes}, dyn_cast<MemRefType>(subviewType).getElementType(),
+      StridedLayoutAttr::get(b.getContext(), ShapedType::kDynamic,
+                             originalStride));
+  auto subview = b.create<memref::SubViewOp>(
+      op.getLoc(), dyn_cast<MemRefType>(subviewType), expandShape.getResult(),
+      offsets, sizes, strides);
+  brgemmOp.getOperand(operandNumber).replaceAllUsesWith(subview);
+}
+
+LogicalResult insertParallelLoop(ForallOp op, unsigned mTileShape,
```

Comment:
I think the current approach disregards other IV users within the loop and can change their iteration space.
Consider this example:
```mlir
func.func @entry(%arg0: memref<8x4x4x4xbf16>, %arg1: memref<4x4x2x4x2xbf16>, %arg2: memref<8x4x4x4xbf16>,
    %arg5: memref<4x4xbf16>) -> memref<4x4xbf16> {
  %c4_i64 = arith.constant 4 : i64
  scf.forall (%arg3, %arg4) in (8, 4) {
    %subview = memref.subview %arg0[%arg3, 0, 0, 0] [1, 4, 4, 4] [1, 1, 1, 1] : memref<8x4x4x4xbf16> to memref<4x4x4xbf16, strided<[16, 4, 1], offset: ?>>
    %subview_0 = memref.subview %arg1[%arg4, 0, 0, 0, 0] [1, 4, 2, 4, 2] [1, 1, 1, 1, 1] : memref<4x4x2x4x2xbf16> to memref<4x2x4x2xbf16, strided<[16, 8, 2, 1], offset: ?>>
    %subview_1 = memref.subview %arg2[%arg3, %arg4, 0, 0] [1, 1, 4, 4] [1, 1, 1, 1] : memref<8x4x4x4xbf16> to memref<4x4xbf16, strided<[4, 1], offset: ?>>
    %0 = xsmm.brgemm.dispatch [4, 4, 4, 4, 4, 4, 16, 16] flags = (vnni_b) data_type = bf16
    xsmm.brgemm(data_type = bf16, %0, %subview, %subview_0, %subview_1, %c4_i64) : (i64, memref<4x4x4xbf16, strided<[16, 4, 1], offset: ?>>, memref<4x2x4x2xbf16, strided<[16, 8, 2, 1], offset: ?>>, memref<4x4xbf16, strided<[4, 1], offset: ?>>, i64) -> ()
    %subview_2 = memref.subview %arg0[%arg3, %arg4, 0, 0] [1, 1, 4, 4] [1, 1, 1, 1] : memref<8x4x4x4xbf16> to memref<4x4xbf16, strided<[4, 1], offset: ?>>
    linalg.add ins(%subview_2, %arg5 
      : memref<4x4xbf16, strided<[4, 1], offset: ?>>, memref<4x4xbf16>)
      outs(%arg5 : memref<4x4xbf16>)
  }
  return %arg5 : memref<4x4xbf16>
}
```

Reply:
Fixed this, working on improving this to use to not use mul-add chains and indexing instead.

Reply:
updated a test